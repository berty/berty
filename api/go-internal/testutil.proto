syntax = "proto3";

package testutil;

option go_package = "berty.tech/berty/go/internal/testutil";

service TestService {
  // use to test unary
  rpc EchoTest(EchoTest.Request) returns (EchoTest.Reply);

  // Use to test stream.
  rpc EchoStreamTest(EchoStreamTest.Request) returns (stream EchoStreamTest.Reply);

  // Use to test duplex stream.
  rpc EchoDuplexTest(stream EchoDuplexTest.Request) returns (stream EchoDuplexTest.Reply);
}

message EchoTest {
  message Request {
    int64 delay = 1;
    string echo = 2;
    bool trigger_error = 3;
  }
  message Reply {
    string echo = 1;
  }
}

message EchoStreamTest {
  message Request {
    int64 delay = 1;
    string echo = 2;
    bool trigger_error = 3;
  }
  message Reply {
    string echo = 1;
  }
}

message EchoDuplexTest {
  message Request {
    string echo = 1;
    bool trigger_error = 2;
  }
  message Reply {
    string echo = 1;
  }
}
