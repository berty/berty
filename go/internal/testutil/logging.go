package testutil

import (
	"flag"
	"os"
	"strings"
	"sync"
	"testing"

	"github.com/stretchr/testify/assert"
	"go.uber.org/zap"

	"berty.tech/berty/v2/go/internal/logutil"
)

const defaultLoggingFilters = "info+:bty.test* error+:*,-ipfs*"

var (
	logFilters     = flag.String("log-filters", defaultLoggingFilters, "log namespaces")
	logFile        = flag.String("log-file", "", "log to file")
	logFormat      = flag.String("log-format", "color", "json, console, color")
	loggerInstance *zap.Logger
	loggerCleanup  func()
	loggerInitOnce sync.Once
)

func Logger(t *testing.T) (*zap.Logger, func()) {
	t.Helper()

	loggerInitOnce.Do(func() {
		if val := os.Getenv("BERTY_LOGFILTERS"); val != "" {
			*logFilters = val
		}
		if val := os.Getenv("BERTY_LOGFILE"); val != "" {
			*logFile = val
		}
		if val := os.Getenv("BERTY_LOGFORMAT"); val != "" {
			*logFormat = val
		}
		*logFilters = strings.ReplaceAll(*logFilters, ":default:", defaultLoggingFilters)

		var err error
		loggerInstance, loggerCleanup, err = logutil.NewLogger(*logFilters, *logFormat, *logFile)
		if !assert.NoError(t, err) {
			loggerInstance = zap.NewNop()
			loggerCleanup = func() {}
		}
	})

	return loggerInstance, loggerCleanup
}
