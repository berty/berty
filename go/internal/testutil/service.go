package testutil

import (
	context "context"
	"testing"
	"time"

	"github.com/stretchr/testify/assert"
	"go.uber.org/zap"
	"google.golang.org/grpc"

	"berty.tech/berty/v2/go/pkg/errcode"
	"berty.tech/weshnet/v2/pkg/grpcutil"
)

type service struct {
	logger *zap.Logger

	UnimplementedTestServiceServer
}

type Options struct {
	Logger *zap.Logger
}

func (o *Options) applyDefault() {
	if o.Logger == nil {
		o.Logger = zap.NewNop()
	}
}

func TestingNewServiceClient(ctx context.Context, t *testing.T, opts *Options) (*grpc.ClientConn, *grpc.Server) {
	t.Helper()

	srv := grpc.NewServer()
	svc := NewService(opts)
	RegisterTestServiceServer(srv, svc)

	l := grpcutil.NewBufListener(2048)

	cc, err := l.NewClientConn(ctx)
	assert.NoError(t, err)

	go func() {
		_ = srv.Serve(l.Listener)
	}()

	t.Cleanup(func() { l.Close() })

	return cc, srv
}

func NewService(opts *Options) TestServiceServer {
	opts.applyDefault()
	return &service{
		logger: opts.Logger,
	}
}

func (svc *service) EchoTest(_ context.Context, req *EchoTest_Request) (*EchoTest_Reply, error) {
	if req.TriggerError {
		return nil, errcode.ErrCode_ErrTestEcho
	}

	if req.Delay > 0 {
		time.Sleep(time.Duration(req.Delay) * time.Millisecond)
	}

	return &EchoTest_Reply{Echo: req.Echo}, nil
}

func (svc *service) EchoStreamTest(req *EchoStreamTest_Request, srv TestService_EchoStreamTestServer) error {
	if req.TriggerError {
		return errcode.ErrCode_ErrTestEcho
	}

	for {
		err := srv.Send(&EchoStreamTest_Reply{Echo: req.Echo})
		if err != nil {
			return errcode.ErrCode_ErrTestEchoSend.Wrap(err)
		}

		time.Sleep(time.Duration(req.Delay) * time.Millisecond)
	}
}

func (svc *service) EchoDuplexTest(srv TestService_EchoDuplexTestServer) error {
	for {
		req, err := srv.Recv()
		if err != nil {
			return errcode.ErrCode_ErrTestEchoRecv.Wrap(err)
		}

		if req.TriggerError {
			return errcode.ErrCode_ErrTestEcho
		}

		err = srv.Send(&EchoDuplexTest_Reply{
			Echo: req.Echo,
		})
		if err != nil {
			return errcode.ErrCode_ErrTestEchoSend.Wrap(err)
		}
	}
}
