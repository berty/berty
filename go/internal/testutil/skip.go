package testutil

import (
	"os"
	"testing"
)

func SkipSlow(t *testing.T) {
	t.Helper()
	if os.Getenv("SKIP_SLOW_TESTS") == "1" {
		t.Skip("slow test skipped")
	}
}
