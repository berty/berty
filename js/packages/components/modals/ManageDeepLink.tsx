import React from 'react'
import { StyleSheet, ActivityIndicator } from 'react-native'
import BlurView from '../shared-components/BlurView'
import InvalidScan from './InvalidScan'
import { Messenger } from '@berty-tech/hooks'
import { useStyles } from '@berty-tech/styles'
import { SafeAreaView } from 'react-native-safe-area-context'
import { SendContactRequest } from './SendContactRequest'
import { ScreenProps } from '@berty-tech/navigation'
import { ManageGroupInvitation } from './ManageGroupInvitation'
import { bufToStr } from '@berty-tech/store/utils'

export const ManageDeepLink: React.FC<ScreenProps.Modals.ManageDeepLink> = ({
	route: { params },
}) => {
	const account = Messenger.useAccount()
	const client = Messenger.useClient()
	const deepLinkStatus = Messenger.useDeepLinkStatus()
	const [{ border }] = useStyles()
	console.log('params', params)
	console.log('ManageDeepLink render:', JSON.stringify(deepLinkStatus, null, 2))
	const dataType = (!params.link && params.type) || 'link'
	let content
	if (!client || !account?.onboarded) {
		content = (
			<InvalidScan
				title={'App not ready!'}
				error={'Please finish the onboarding before using links or QR codes.'}
			/>
		)
	} else if (
		!deepLinkStatus ||
		(deepLinkStatus.link !== params.value && deepLinkStatus.link !== params.link)
	) {
		content = <ActivityIndicator size='large' />
	} else if (deepLinkStatus.error !== undefined) {
		let title
		if (dataType === 'link') {
			title = 'Invalid link!'
		} else if (dataType === 'qr') {
			title = 'Invalid QR code!'
		} else {
			title = 'Error!'
		}
		content = <InvalidScan title={title} error={deepLinkStatus.error} />
	} else if (deepLinkStatus.kind === 'group') {
		const { bertyGroup } = deepLinkStatus.parsedData || {}
		if (!(bertyGroup?.group?.publicKey && bertyGroup?.displayName)) {
			content = (
				<InvalidScan
					title={'Error joining group!'}
					error='This invitation contains incomplete data. ☹️'
				/>
			)
		} else {
			content = (
				<ManageGroupInvitation
					groupPk={bufToStr(bertyGroup.group.publicKey)}
					displayName={bertyGroup.displayName}
				/>
			)
		}
	} else if (deepLinkStatus.kind === 'contact') {
		content = <SendContactRequest type={dataType} />
	}
	return (
		<>
			<BlurView style={[StyleSheet.absoluteFill]} blurType='light' />
			<SafeAreaView style={[border.shadow.huge]}>{content}</SafeAreaView>
		</>
	)
}
