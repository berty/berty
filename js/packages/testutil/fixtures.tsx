import beapi from '@berty-tech/api'

const empty: beapi.messenger.IMedia = {
	metadataBytes: beapi.messenger.MediaMetadata.encode({}).finish(),
}

const media = {
	empty,
}

export default {
	media,
}
