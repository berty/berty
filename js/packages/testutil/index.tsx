import React from 'react'
import { NavigationContainer } from '@react-navigation/native'
import { IconRegistry } from '@ui-kitten/components'
import { EvaIconsPack } from '@ui-kitten/eva-icons'
import { SafeAreaProvider } from 'react-native-safe-area-context'

import { Provider as ThemeProvider } from '@berty-tech/components/theme'
import { FeatherIconsPack } from '@berty-tech/messenger-app/feather-icons'
import { CustomIconsPack } from '@berty-tech/messenger-app/custom-icons'
import { Provider as StyleProvider } from '@berty-tech/styles'

export const TestProvider: React.FC = ({ children }) => (
	<SafeAreaProvider
		initialMetrics={{
			frame: { x: 0, y: 0, width: 0, height: 0 },
			insets: { top: 0, left: 0, right: 0, bottom: 0 },
		}}
	>
		<StyleProvider>
			<IconRegistry icons={[EvaIconsPack, FeatherIconsPack, CustomIconsPack]} />
			<ThemeProvider>
				<NavigationContainer>{children}</NavigationContainer>
			</ThemeProvider>
		</StyleProvider>
	</SafeAreaProvider>
)

import _fixtures from './fixtures'
export const fixtures = _fixtures
