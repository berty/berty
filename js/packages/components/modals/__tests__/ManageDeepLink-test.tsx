import React from 'react'
import renderer from 'react-test-renderer'

import { TestProvider } from '@berty-tech/testutil'

import { ManageDeepLink } from '../ManageDeepLink'

test('ManageDeepLink modal renders correctly', () => {
	const props: React.ComponentProps<typeof ManageDeepLink> = {
		route: {
			params: {
				type: 'link',
				value:
					'https://berty.tech/id#contact/oZBLFWJdsE73yWa2BjceiUtKCfxpxVBuH8h6r3sJXDrgqaHyR1ACgLpWEZTMDD8DzGsSo679UjknhcuXGiqP6cgU66R5BTh/name=Alice',
			},
		},
	}
	const tree = renderer
		.create(
			<TestProvider>
				<ManageDeepLink {...props} />
			</TestProvider>,
		)
		.toJSON()
	expect(tree).toMatchSnapshot()
})
