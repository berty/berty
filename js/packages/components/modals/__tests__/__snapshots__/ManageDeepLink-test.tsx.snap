// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ManageDeepLink modal renders correctly 1`] = `
<RNCSafeAreaProvider
  onInsetsChange={[Function]}
  style={
    Array [
      Object {
        "flex": 1,
      },
      undefined,
    ]
  }
>
  <View
    style={
      Object {
        "flex": 1,
      }
    }
  >
    <BlurView
      blurAmount={10}
      blurType="light"
      style={
        Array [
          Object {
            "backgroundColor": "transparent",
          },
          Array [
            Object {
              "bottom": 0,
              "left": 0,
              "position": "absolute",
              "right": 0,
              "top": 0,
            },
          ],
        ]
      }
    />
    <RNCSafeAreaView
      style={
        Array [
          Object {
            "shadowOffset": Object {
              "height": 6.5,
              "width": 0,
            },
            "shadowOpacity": 0.2,
            "shadowRadius": 13,
          },
          Object {
            "shadowColor": "#000000",
          },
        ]
      }
    >
      <ActivityIndicator
        animating={true}
        color="#999999"
        hidesWhenStopped={true}
        size="large"
      />
    </RNCSafeAreaView>
  </View>
</RNCSafeAreaProvider>
`;
